Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,357 @@
+#include <Kernel/FileSystem/FATFileSystem.h>
+#include <Kernel/Memory/Region.h>
+
+namespace Kernel {
+
+ErrorOr<NonnullLockRefPtr<FileSystem>> FATFS::try_create(OpenFileDescription& file_description)
+{
+    return TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) FATFS(file_description)));
+}
+
+FATFS::FATFS(OpenFileDescription& file_description)
+    : BlockBasedFileSystem(file_description)
+{
+}
+
+ErrorOr<void> FATFS::initialize()
+{
+    MutexLocker locker(m_lock);
+
+    m_boot_record = TRY(KBuffer::try_create_with_size("FATFS: Boot Record"sv, m_logical_block_size));
+    auto boot_record_buffer = UserOrKernelBuffer::for_kernel_buffer(m_boot_record->data());
+    TRY(raw_read(0, boot_record_buffer));
+
+    if constexpr (FAT_DEBUG) {
+        dbgln("FATFS: oem_identifier: {}", boot_record()->oem_identifier);
+        dbgln("FATFS: bytes_per_sector: {}", boot_record()->bytes_per_sector);
+        dbgln("FATFS: sectors_per_cluster: {}", boot_record()->sectors_per_cluster);
+        dbgln("FATFS: reserved_sector_count: {}", boot_record()->reserved_sector_count);
+        dbgln("FATFS: fat_count: {}", boot_record()->fat_count);
+        dbgln("FATFS: root_directory_entry_count: {}", boot_record()->root_directory_entry_count);
+        dbgln("FATFS: media_descriptor_type: {}", boot_record()->media_descriptor_type);
+        dbgln("FATFS: sectors_per_track: {}", boot_record()->sectors_per_track);
+        dbgln("FATFS: head_count: {}", boot_record()->head_count);
+        dbgln("FATFS: hidden_sector_count: {}", boot_record()->hidden_sector_count);
+        dbgln("FATFS: sector_count: {}", boot_record()->sector_count);
+        dbgln("FATFS: sectors_per_fat: {}", boot_record()->sectors_per_fat);
+        dbgln("FATFS: flags: {}", boot_record()->flags);
+        dbgln("FATFS: fat_version: {}", boot_record()->fat_version);
+        dbgln("FATFS: root_directory_cluster: {}", boot_record()->root_directory_cluster);
+        dbgln("FATFS: fs_info_sector: {}", boot_record()->fs_info_sector);
+        dbgln("FATFS: backup_boot_sector: {}", boot_record()->backup_boot_sector);
+        dbgln("FATFS: drive_number: {}", boot_record()->drive_number);
+        dbgln("FATFS: volume_id: {}", boot_record()->volume_id);
+    }
+
+    if (boot_record()->signature != signature_1 && boot_record()->signature != signature_2) {
+        dbgln("FATFS: Invalid signature");
+        return EINVAL;
+    }
+
+    m_logical_block_size = boot_record()->bytes_per_sector;
+    set_block_size(m_logical_block_size);
+
+    u32 root_directory_sectors = ((boot_record()->root_directory_entry_count * sizeof(FATEntry)) + (m_logical_block_size - 1)) / m_logical_block_size;
+    m_first_data_sector = boot_record()->reserved_sector_count + (boot_record()->fat_count * boot_record()->sectors_per_fat) + root_directory_sectors;
+
+    TRY(BlockBasedFileSystem::initialize());
+
+    FATEntry root_entry {};
+    root_entry.first_cluster_low = FATInode::root_inode_index;
```

Comment:
Yes they do, that's a part of the `FATEntry` struct which we do not have control of, because it's a part of the spec. I have added a helper function to read it.

Reply:
I ended up getting rid of the `root_inode_index` constant and replaced it with the value that would be actually used later, now both `first_cluster_low` and `first_cluster_high` are being written when creating the root inode.